Automattic/woocommerce-payments
# Add P24 as an UPE payment method

Fixes #2651

#### Changes proposed in this Pull Request

TBD

Add Przelewy24 (P24) as an UPE payment method to WooCommerce Payments.


This PR adds the P24 payment method as an UPE provider to WooCommerce Payments. P24 is the most popular payment method in Poland, it supports PLN and EUR as payment currencies, also from their website, they support:

- card transaction processing on the merchant's website, the option of adapting the card form to the website/store layout,
- One Click Payment - payment with a single click - with no need for the buyer to create an account in the payment service provider's system,
- supporting transactions in 50 currencies with no currency conversion fees - 1 to 1 payment,
- DCC (Dynamic Currency Conversion) - this solution allows customers to make transactions in foreign currencies. A foreign card-holder may choose the currency for a given transaction but is charged the amount in his native currency. When this type of payment is made, the merchant receives a bonus - a refund of up to 1% of the transaction value,
- recurring payments which guarantee timely settlements with merchants who offer subscriptions, passes, etc.,
- card payments over the phone - IVR
- transfer of risk in case of fraud.





After the implementation, these screens now contain P24 elements:

##### Add Payment Method Modal
<img width="571" alt="Add Payment Methods" src="https://user-images.githubusercontent.com/3295/128630812-0b78e8cf-df74-4bea-8d78-c0692e13d1ba.png">

##### Enabled Payment Methods List
<img width="677" alt="Enabled Payment Methods" src="https://user-images.githubusercontent.com/3295/128630814-0bc56d0b-d4ee-4877-9088-9ebc2a5ac7c7.png">

##### Payment Screen
<img width="264" alt="Payment Screen" src="https://user-images.githubusercontent.com/3295/128630816-5c02f5f4-6a98-4e5c-9772-90ebe47de2fe.png">

##### Transactions List
<img width="1483" alt="Transactions List" src="https://user-images.githubusercontent.com/3295/128630820-6d795c30-2544-449e-bd7b-86ac49e7c1af.png">

##### Transaction Details
<img width="849" alt="Transaction Details" src="https://user-images.githubusercontent.com/3295/128630819-2b7d4070-a3ce-42a1-a89f-98b1c53e2171.png">

##### Refunds List
<img width="1499" alt="Refund" src="https://user-images.githubusercontent.com/3295/128630818-e3e8e6c9-9bd7-4530-ada3-6d06a83845c8.png">

##### Refund Details
<img width="527" alt="Refund Details" src="https://user-images.githubusercontent.com/3295/128630817-632b66dc-5887-447c-ba0d-51b83dec8b04.png">

#### Testing instructions





* Go to **WooCommerce > Settings > Payments** And on this screen, find **Payment Methods** metabox (ensure you have grouped settings enabled)
* Check if you can see the P24 icon on the right of the **Add Payment Method** button.
* Click the **Add Payment Method** button, select **Przelewy24 (P24)** and click **Add selected**.
* Check if **Przelewy24 (P24)** is added to the **Payment Methods** list.
* Go to your store and select your customer currency to EUR. Or if you don't have customer multi currency, set your store currency to EUR from **WooCommerce > Settings > General** page.
* Add some items to the cart, and proceed to checkout. 
* On the checkout page, select **Use a different payment method** radio button, and wait for the UPE iframe to render.
* Check if you can see Przelewy24 (P24) on the available payment methods list in the iframe.
* Select P24 and on the new form below, select a bank.
* Click **Place Order** button to complete the payment.
* This should redirect you to Stripe's P24 test transaction page. 

![image](https://user-images.githubusercontent.com/3295/128631115-d700b994-8057-4310-a287-7f39a34582fa.png)

* On that page, you have two options:
    - Click **Fail test payment** and check if a Payment error is displayed after that on the checkout page.
    - Click **Authorize test payment** and check if you get redirected to the order received page. 

* After you completed the payment, go to admin page **Payments > Transactions** (you may need to sync the transactions via CLI, or have server hook listener open to update it) and check if you can see the transaction with the P24 icon on it. 
* Click that row and check the details if the details are correct.

**Note:** Whatever you select as a bank on P24 payment page, it returns a hardcoded response as the success message with this data:

```json
"payment_method_details": {
    "p24": {
        "bank": "ing",
        "reference": "P24-N01-101-101 STRIPE POLAND",
        "verified_name": "Jenny Rosen"
    },
    "type": "p24"
},
```

-------------------
